internal/xdscache/v3/endpointslicetranslator.go

```diff
@@ -0,0 +1,458 @@
+	var healthCheckPort int32
+
+	for _, enpointSlice := range endpointSliceMap {
+		sort.Slice(enpointSlice.Endpoints, func(i, j int) bool {
+			return enpointSlice.Endpoints[i].Addresses[0] < enpointSlice.Endpoints[j].Addresses[0]
+		})
+
+		for _, endpoint := range enpointSlice.Endpoints {
+			// Skip if the endpointSlice is not marked as ready.
+			if endpoint.Conditions.Ready != nil && !*endpoint.Conditions.Ready {
+				continue
+			}
+
+			// Range over each port. We want the resultant endpoints to be a
+			// a cartesian product MxN where M are the endpoints and N are the ports.
+			for _, p := range enpointSlice.Ports {
+				// Nil check for the port.
+				if p.Port == nil {
+					continue
+				}
+
+				if p.Protocol == nil || (healthPort.Protocol != *p.Protocol || port.Protocol != *p.Protocol) && *p.Protocol != v1.ProtocolTCP {
+					continue
+				}
+
+				// Set healthCheckPort only when port and healthPort are different.
+				if p.Name != nil && (healthPort.Name != "" && healthPort.Name == *p.Name && port.Name != healthPort.Name) {
+					healthCheckPort = *p.Port
+				}
+
+				// Match by port name.
+				if port.Name != "" && p.Name != nil && port.Name != *p.Name {
+					continue
+				}
+
+				// we can safely take the first element here.
+				// Refer to: https://issue.k8s.io/106267
+				addr := envoy_v3.SocketAddress(endpoint.Addresses[0], int(*p.Port))
+
+				// as per note on https://kubernetes.io/docs/concepts/services-networking/endpoint-slices/
+				// Clients of the EndpointSlice API must iterate through all the existing EndpointSlices associated to
+				// a Service and build a complete list of unique network endpoints. It is important to mention that
+				// endpoints may be duplicated in different EndpointSlices.
+				// Hence, we need to ensure that the endpoints we add to []*LoadBalancingEndpoint aren't duplicated.
+				endpointKey := fmt.Sprintf("%s:%d", endpoint.Addresses[0], *p.Port)
+				if _, exists := uniqueEndpoints[endpointKey]; !exists {
+					lb = append(lb, envoy_v3.LBEndpoint(addr))
+					uniqueEndpoints[endpointKey] = struct{}{}
+				}
+			}
+		}
+	}
+
+	if healthCheckPort > 0 {
+		for _, lbEndpoint := range lb {
+			lbEndpoint.GetEndpoint().HealthCheckConfig = envoy_v3.HealthCheckConfig(healthCheckPort)
+		}
+	}
+
+	return lb
+}
+
+// EndpointSliceCache is a cache of EndpointSlice and ServiceCluster objects.
+type EndpointSliceCache struct {
+	mu sync.Mutex // Protects all fields.
+
+	// Slice of stale clusters. A stale cluster is one that
+	// needs to be recalculated. Clusters can be added to the stale
+	// slice due to changes in Endpoints or due to a DAG rebuild.
+	stale []*dag.ServiceCluster
+
+	// Index of ServiceClusters. ServiceClusters are indexed
+	// by the name of their Kubernetes Services. This makes it
+	// easy to determine which Endpoints affect which ServiceCluster.
+	services map[types.NamespacedName][]*dag.ServiceCluster
+
+	// Cache of endpointsSlices, indexed by Namespaced name.
```

Comment:
The key is the NamespacedName of the _Service_, not the EndpointSlice -- let's note that here.